### PR TITLE
fix: add repo health job workflow

### PR DESCRIPTION
**Pull request by UsamaSadiq**
### Description 
- Fixes the issue https://github.com/edx/edx-arch-experiments/issues/66
- Added a reusable workflow to run the repo health job on given organisation's repos.
- Added template workflow for organisations to trigger & use the repo-health-job reusable workflow.
- Some constraints will be removed after completing the testing on the PR. 

### Testing
- Reusable workflow has been tested in the fork both ways: 
1. By being triggered from a workflow under same organisation
2. By being triggered from a workflow under different organisation

### CODE REVIEW

File: .github/workflows/repo-health-job.yml
Comment (jmbowman) [inline .github/workflows/repo-health-job.yml:84]: This should be a parameter, since anybody other than 2U will be storing their data in a different repo.

File: workflow-templates/repo-health-workflow.yml
Comment (jmbowman) [inline workflow-templates/repo-health-workflow.yml:22]: It feels like all of this stuff for getting data from or updating in Google Sheets should be in the 2U private workflow call, rather than the shared workflow.

File: scripts/repo-health-script.sh
Comment (nedbat) [reply scripts/repo-health-script.sh:22]: Also, this ignored repo is a temporary private repo in the edx GItHub org, so shouldn't be hard-coded here.

File: .github/workflows/repo-health-job.yml
Comment (nedbat) [inline .github/workflows/repo-health-job.yml:101]: We've checked out this repo, so this line can run the local file, I think?

File: scripts/repo-health-script.sh
Comment (nedbat) [reply scripts/repo-health-script.sh:57]: Why do we skip edx-repo-health? Don't we want to check it also?

File: scripts/repo-health-script.sh
Comment (nedbat) [inline scripts/repo-health-script.sh:100]: Lines 100 and 97 are really long, and (intentionally) identical.  We should put the command in a variable, with backslashes for readability.

File: scripts/repo-health-script.sh
Comment (nedbat) [inline scripts/repo-health-script.sh:113]: Don't forget to remove this before merging! :)

File: .github/workflows/repo-health-job.yml
Comment (UsamaSadiq) [reply .github/workflows/repo-health-job.yml:58]: Updated as per the suggestion.

File: workflow-templates/repo-health-workflow.yml
Comment (UsamaSadiq) [reply workflow-templates/repo-health-workflow.yml:28]: This logic for getting and updating data in sheets was copied to provide this feature to other orgs as well. Since the values of these parameters will be coming from secrets, other orgs can just save their own credentials in secrets to avail the feature of populating and maintaining their own repo health sheet. 

If we want to keep this feature only within 2U, then we'll need to create a seperate workflow for 2U org. 

Right now, the whole shared workflow/script is exactly the same as the jenkins job and different orgs can provide their own paths/secrets to store their data on their desired paths/repos.

File: .github/workflows/repo-health-job.yml
Comment (UsamaSadiq) [reply .github/workflows/repo-health-job.yml:112]: When the reusable/shared workflow was triggered from a workflow under different organisation, it only copied the workflow file and couldn't access the script file directly. 
We had two options of either going with creating a composite workflow or downloading the script to be used so we went with the later approach since it was simple and easier to understand.

File: scripts/repo-health-script.sh
Comment (UsamaSadiq) [reply scripts/repo-health-script.sh]: Updated the lines as per the suggestion.

File: scripts/repo-health-script.sh
Comment (UsamaSadiq) [reply scripts/repo-health-script.sh]: Removed this loop and other variables added to enable quick script execution for testing.

**Comment by UsamaSadiq** [comment]
@nedbat @jmbowman 
I tried re-requesting review from both of you but for some reason, GitHub wants me to only ask for review from one person at a time 😅 and removes other one automatically.

File: workflow-templates/repo-health-workflow.yml
Comment (jmbowman) [reply workflow-templates/repo-health-workflow.yml:28]: I guess this is ok as long as the Google Sheet updating is gracefully skipped if no data is provided.  I want to make sure that having a Google Workplace account isn't a prerequisite for running the repo health workflow.

File: .github/workflows/repo-health-job.yml
Comment (jmbowman) [reply .github/workflows/repo-health-job.yml:112]: If we never call that script directly from this repo, would it make sense to move it to `edx-repo-health` which is checked out into a local directory earlier in the workflow?

File: workflow-templates/repo-health-workflow.yml
Comment (UsamaSadiq) [reply workflow-templates/repo-health-workflow.yml:28]: We've tried it when testing the workflow from the forked repo, if no google sheet credentials are provided and the workflow is called with default values, then the step for updating google sheet is automatically skipped.

File: scripts/repo-health-script.sh
Comment (nedbat) [inline scripts/repo-health-script.sh:102]: Isn't this going to run the command just once, and use the same result on both lines 103 and 106?  It won't run the command twice, which is what you want.

Also, please indent this so that it look like one continued like rather than a series of commands that start with "REPO", "--repo", etc.

File: scripts/repo-health-script.sh
Comment (nedbat) [inline scripts/repo-health-script.sh:45]: Please use one block style consistently.  Most favored seems to be:
```
while IFS= read -r line; do
    ...
```

File: scripts/repo-health-script.sh
Comment (nedbat) [inline scripts/repo-health-script.sh:79]: This is another block style to be made uniform.

File: scripts/repo-health-script.sh
Comment (nedbat) [inline scripts/repo-health-script.sh:122]: Backquotes are old-style, can this be done with `$()` instead?

File: scripts/repo-health-script.sh
Comment (nedbat) [inline scripts/repo-health-script.sh:22]: Long lines should be split across lines with backslashes so they can be read more easily.

File: scripts/repo-health-script.sh
Comment (nedbat) [inline scripts/repo-health-script.sh:48]: Are we indenting with two spaces or four?  I don't know if we have a consistent style across the org, but we can use just one within a file.  Four spaces would be best.

File: scripts/repo-health-script.sh
Comment (nedbat) [inline scripts/repo-health-script.sh:69]: Is this `||` another way to write a conditional?  It would be easier to understand if it were written more explicitly.

**Review by nedbat — CHANGES_REQUESTED** [review]
Thanks for taking care of my earlier concerns.  The repo-health-script.sh file is long for a shell script.  I feel better about it if it was more uniform: I counted at least four different styles of if-statements and indentation.

**Comment by UsamaSadiq** [comment]
Hi @nedbat as per Jeremy's suggestion, I'll be moving the `repo-health-job.yml` script into the `edx-repo-health` repo so we won't have to clone it seperately to execute in the workflow. 
I'll address your concerns about the shell script in the `edx-repo-health` PR to add the shell script. Most of these points were also noticed by our team before creating this PR but since this script was already being used in Jenkins like this, we wanted to do the minimal chansges in the script to make it work with our workflows.

**Comment by nedbat** [comment]
> Hi @nedbat as per Jeremy's suggestion, I'll be moving the `repo-health-job.yml` script into the `edx-repo-health` repo so we won't have to clone it seperately to execute in the workflow.
> I'll address your concerns about the shell script in the `edx-repo-health` PR to add the shell script. Most of these points were also noticed by our team before creating this PR but since this script was already being used in Jenkins like this, we wanted to do the minimal chansges in the script to make it work with our workflows.

I guess you mean those changes will be in a new pull request?  Then let's go ahead with this one.

**Review by nedbat — APPROVED** [review]

**Review by jmbowman — APPROVED** [review]
Looks good to me once the linked PR is merged.

**Comment by UsamaSadiq** [comment]
@jmbowman 
Could you merge this PR if you have merge access in the repo? or should I create a tcril issue for this to get this PR merged? 

After this PR is merged:
- [ ] A PR will be created to create the trigger workflow in the 2U org which will use this reusable workflow to populate the existing google sheet. 
- [ ] An SRE ticket will be created to copy the existing credentials from Jenkins env to edx org secrets which will be used by the workflow.
- [ ] An announcement will be made for the open edx community to inform about this tool and list down the steps needed for the other orgs to be able to use this for their reops.

Question: Which will be the best place in the edx org to host the trigger workflow for repo health job? `edx/.github` or `repo-tools`?

**Comment by jmbowman** [comment]
I don't have merge rights, so that'll need to be a tCRIL issue.  And I'd suggest `edx/repo-health-data` for the 2U version of the job, to avoid leaking information about private repositories in public workflow logs.

**Pull request merged**
